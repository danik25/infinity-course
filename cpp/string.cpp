/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}