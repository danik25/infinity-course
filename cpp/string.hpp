/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */