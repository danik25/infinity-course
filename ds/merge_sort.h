/****************************************************************

*                      Merge Sort

* 						=====================

* File Name: merge_sort.h

* Related files: merge_sort.c   merge_sort_test.c

* #Version: V 1.1

* Writer: Dani Benjamin      

* Reviewer: Tom Valershtein     

* Last update: 04.06.2019

****************************************************************/

#ifndef _MERGE_SORT_H_
#define _MERGE_SORT_H_

#include <stddef.h>

enum STATUS {SUCCESS, MEM_ALLOC_FAIL};


typedef int (*is_before_t)(const void *data1, const void *data2);


int MergeSort(void *array,
    		  size_t arr_size,
              size_t element_size,
              is_before_t IsBefore);

#endif