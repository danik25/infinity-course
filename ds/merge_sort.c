/****************************************************************

*                      Merge Sort

*                       =====================

* File Name: merge_sort.h

* Related files: merge_sort.c   merge_sort_test.c

* #Version: V 1.1

* Writer: Dani Benjamin      

* Reviewer: Tom Valershtein     

* Last update: 04.06.2019

****************************************************************/

#include <stdlib.h> /* malloc */
#include <string.h> /* memcpy */
#include <assert.h> /* assert */
#include "merge_sort.h"

static void RecMergeSort(void *arr, void* tmp_res, size_t start, size_t end,
                                is_before_t IsBefore, size_t element_size);
static void Merge(void *arr, void *res, size_t start, size_t mid, size_t end, 
                                is_before_t IsBefore, size_t element_size);

#define START_OF_ARRAY 0

int MergeSort(void *array, size_t arr_size, 
                    size_t element_size, is_before_t IsBefore)
{
    void* handle = (void*)malloc(element_size*arr_size);

    assert(array);
    assert(IsBefore);

    if(NULL == handle)
    {
        return (MEM_ALLOC_FAIL);
    }
    RecMergeSort(array, handle, START_OF_ARRAY, arr_size - 1, IsBefore, element_size);
   
    free(handle);

    return(SUCCESS);
}

static void RecMergeSort(void *arr, void* tmp_res, size_t start, size_t end,
                                is_before_t IsBefore, size_t element_size)
{
    size_t mid = 0;

    assert(arr);
    assert(tmp_res);
    assert(IsBefore);

    if(start >= end)
    {
        return ;
    }
    
    mid = (start + end) / 2;
    RecMergeSort(arr, tmp_res, start, mid, IsBefore, element_size);
    RecMergeSort(arr, tmp_res, mid + 1, end, IsBefore, element_size);
    Merge(arr, tmp_res, start, mid, end, IsBefore, element_size);
       
}

static void Merge(void *arr, void *res, size_t start, size_t mid, size_t end, 
                                is_before_t IsBefore, size_t element_size)
{
    size_t left = start;
    size_t right = mid + 1;
    char res_index = start;
    char *arr_copy = NULL;
    char *res_copy = NULL;

    assert(arr);
    assert(res);
    assert(IsBefore);

    arr_copy = (char*)arr;
    res_copy = (char*)res;

    while(left <= mid && right <= end)
    {
        if(IsBefore((arr_copy + right * element_size), 
                        (arr_copy + left * element_size)))
        {
            memcpy(res_copy + res_index * element_size, 
                        arr_copy + right * element_size, element_size);
            ++right;
        }

        else
        {
            memcpy(res_copy + res_index * element_size, 
                        arr_copy + left * element_size, element_size);
            ++left;
        }
        ++res_index;
    }

    
    memcpy(res_copy + res_index * element_size, 
            arr_copy + right * element_size, (end - right + 1) * element_size);
   
    memcpy(res_copy + res_index * element_size, 
            arr_copy + left * element_size, (mid - left + 1) * element_size); 

    memcpy((char*)arr + start * element_size, 
            (char*)res_copy + start * element_size, (end - start + 1) * element_size);  
}