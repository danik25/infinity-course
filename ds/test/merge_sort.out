/home/xyz/git/dani-benjamin/ds/merge_sort.out